### ISSUE TITLE
Why don't you sample Alphas and Betas using Gumbel-Softmax?

### ISSUE THREAD

**Issue by maaft** [TAMU-VITA/FasterSeg#13]
In your model_search.py code you simply use `F.softmax(...)` on alphas and betas to learn the super-cell and down-sampling connections (line 275):

```python
        alphas0 = F.softmax(getattr(self, self._arch_names[self.arch_idx]["alphas"][0]), dim=-1)
        alphas1 = F.softmax(getattr(self, self._arch_names[self.arch_idx]["alphas"][1]), dim=-1)
        alphas2 = F.softmax(getattr(self, self._arch_names[self.arch_idx]["alphas"][2]), dim=-1)
        alphas = [alphas0, alphas1, alphas2]
        betas1 = F.softmax(getattr(self, self._arch_names[self.arch_idx]["betas"][0]), dim=-1)
        betas2 = F.softmax(getattr(self, self._arch_names[self.arch_idx]["betas"][1]), dim=-1)
        betas = [None, betas1, betas2]`
```
I was wondering why you didn't use hard Gumbel-Softmax sampling here (like you did with the ratios)?

I think that using plain softmax is actually very dangerous as the optimizer learns to output optimal weighted **combinations** of feature-maps from all available operations.

Furthermore, I observed in a small Experiment using only 2 Operations (Conv and Skip) that using plaing softmax didn't converge at all but Gumbel-Softmax-Sampling leads to good `alphas`.

What do you think? Did I miss something?

**Comment by chenwydj** [comment]
Hi @maaft! Thank you for your interest in our work!

As an example, you may want to take a look at Figure 2 in https://openreview.net/forum?id=SJx9ngStPH. It seems the Gumbel softmax ("GDAS") may lead to worse one-shot validation error, while at the same time achieve similar test regret (i.e., the accuracy of derived subnetworks).

For now, it is hard for me to say if using Gumbel softmax is always a good idea. However, I would guess stabilizing feature maps in the supernet should lead to a stable search process, which could be vital for dense prediction tasks like segmentation. Gumbel softmax may include larger dynamics during the search.

**Comment by maaft** [comment]
Thank you for your answer!

To clarify, alphas and betas are learned using softmax during training and the network for evaluation is just picked using argmax, correct?

I.e.:
```python
def train(...):
   alphas = F.softmax(self.alphas, dim=-1)
   ...

def eval(...):
    idx = self.alphas.argmax(dim=-1, keepdim=True)
    alphas = torch.zeros_like(alphas)
    alphas.scatter_(1, idx, 1)

```

Do you mean by "stabilizing feature maps" adding a loss that forces each cell to produce a similar output to other cells of the same supercell?

**Comment by chenwydj** [comment]
Yes -- the operator with the largest alpha value is selected for train-from-scratch.

By "stabilizing feature maps" I mean reducing the variations of feature maps during the search. A large variation could come from different feature maps produced by different operators sampled by the one-hot Gumble softmax. Instead, the feature map will be less variant if it is the weighted sum of all operators' outputs.

**Comment by maaft** [comment]
Thanks!
Does this also apply to betas? If yes, how can the searched architecture have two ingoing connections from previous cells? (E.g. Cell 7 from downsampling-level 16)

![image](https://user-images.githubusercontent.com/2518993/75244570-4a512c00-57cc-11ea-9f9a-dbd603dc4aa5.png)

Or do you use some kind of thresholds to determine if a connection is disabled or not? And if yes, what is the threshold you are using?

**Comment by chenwydj** [comment]
At index #7 there are actually two cells. See the last paragraph in our section 3.4 (page 6 bottom):

"It is worth noting that, the multi-resolution branches will share both cell weights and feature maps if their cells are of the same operator type, spatial resolution, and expansion ratio. This design contributes to a faster network. **Once cells in branches diverge, the sharing between the branches will be stopped and they become individual branches**."

This means starting from index #5, two branches will never share anything. They have their separate convolutional layers.

For your Gumbel Softmax question: I never tried to sample per-image Gumbel noise. If you did not apply temperature annealing in the Gumbel Softmax, then maybe you could try: 1) use smaller architecture learning rate; 2) use larger temperature for Gumbel Softmax.

**Comment by maaft** [comment]
I've read that passage but I don't really understand it as your code is suggesting something else:

In `model_search.py` you initialize your `Network_Multi_Path` with all cells of all branches and I don't see shared parameters anywhere. Could you please clarify where the sharing takes place?

Thank you for being so patient and answering my questions! :)

**Comment by chenwydj** [comment]
We implement this in our derived architectures.

1. At each layer, we say two branches are "connected" (i.e. they share both cell weight and feature maps) only if: 1) they output the same scale, 2) they choose the same operator, 3) they choose the same width: https://github.com/TAMU-VITA/FasterSeg/blob/master/train/model_seg.py#L257
2. We merge connected branches into a group: https://github.com/TAMU-VITA/FasterSeg/blob/master/train/model_seg.py#L261
3. Branches belong the same group will share the same cell, otherwise they will not share either [cells](https://github.com/TAMU-VITA/FasterSeg/blob/master/train/model_seg.py#L293) or [feature maps](https://github.com/TAMU-VITA/FasterSeg/blob/master/train/model_seg.py#L345).

**Comment by maaft** [comment]
Okay, so the sharing is only done after the model search is finished.

Another thing I noticed is that in your paper you use the betas on the input of each cell:
![image](https://user-images.githubusercontent.com/2518993/75525843-e0759400-5a10-11ea-9b3b-c43342b48bb4.png)

But in your implementation you use the betas on the cell's outputs:
https://github.com/TAMU-VITA/FasterSeg/blob/bb52a004ff83f64d3dd8989104234fdb862d1cc5/search/model_search.py#L330

I guess that both approaches are somehow equivalent in the end but I was wondering why you decided to do the weighting on the output of the cell?

Afaik, input weighting as per eq. (1) would spare us one forward pass per cell and therefore highly needed memory.

**Comment by maaft** [comment]
Sorry to bother you again, but could you kindly answer my last question?

If we really are able to multiply betas on the cell's input we would spare lots of memory and could use larger batch-sizes.

**Comment by maaft** [comment]
Also, to come back to my original question:
If you train a **softmax-weighted** sum of all all feature-maps for each super-cell, how do you decide that the architecture sampled from epoch `N+1` is actually better than the architecture from epoch `N` ? 

The problem that I have observed is, that even while the architecture might have gotten better, the argmax-derived model is performing worse (i.e. lower IoU) and I think that this mainly comes from softmax-training.

**Comment by chenwydj** [comment]
> Sorry to bother you again, but could you kindly answer my last question?
> 
> If we really are able to multiply betas on the cell's input we would spare lots of memory and could use larger batch-sizes.

I think current Eq. 1 is actually doing input weighting, as the weighted output will be the input into the next cell.

Without using Gumbel Softmax, I doubt we could save memory here, since both two outputs need to be calculated in the computation graph.

**Comment by chenwydj** [comment]
> Also, to come back to my original question:
> If you train a **softmax-weighted** sum of all all feature-maps for each super-cell, how do you decide that the architecture sampled from epoch `N+1` is actually better than the architecture from epoch `N` ?
> 
> The problem that I have observed is, that even while the architecture might have gotten better, the argmax-derived model is performing worse (i.e. lower IoU) and I think that this mainly comes from softmax-training.
> 
> Here you see the validation DiceScore of the argmax-derived model. Red is with hard-gumbel-softmax sampling, blue is with plain-softmax training:
> 
> ![image](https://user-images.githubusercontent.com/2518993/75785025-65d7ac00-5d63-11ea-92ed-d990eb3711b6.png)

For this question, [this paper](https://openreview.net/pdf?id=SJx9ngStPH) may be helpful to you.

In their Figure 2, solid lines are the performance of derived subnetworks, and dashed lines are the performance of the supernet. We can see that the solid line (test error) is not always getting lower along the search process. I think this is one important challenge in NAS community to be solved, especially for weight-sharing NAS methods.

**Comment by maaft** [comment]
> I think current Eq. 1 is actually doing input weighting, as the weighted output will be the input into the next cell.

Yes, this is what I wrote and understand. But in your code you do output weighting and I don't understand why?

See:
https://github.com/TAMU-VITA/FasterSeg/blob/bb52a004ff83f64d3dd8989104234fdb862d1cc5/search/model_search.py#L331

**Comment by chenwydj** [comment]
I mean output reweighting at `cell_i` _is_ actually doing input reweight for `cell_i+1`.

**Comment by maaft** [comment]
Okay, I will continue to use input-weighting then as I spare one forward pass through my cell (faster search time, lower memory consumption).
